File: CI1056 - Algoritmos 2/Binário/BuscaBinária.c
#include <stdio.h>

#define TAM 200

// Função para ler os elementos do vetor
void lerVetor(int n, int v[]) 
{

    for (int i = 0; i < n; i++) 
    {
        scanf("%d", &v[i]);
    }
}

// Função para realizar a busca binária
int PesqBin(int n, int elemento, int v[]) {
    int inicio = 0;
    int fim = n - 1;
    int meio;

    while (inicio <= fim) 
    {
        meio = (inicio + fim)/2;
        if (elemento > v[meio]) 
        {
            inicio = meio + 1;
        } 
        else if (elemento < v[meio]) 
        {
            fim = meio - 1;
        } 
        else 
        {
            return meio; // Índice do elemento encontrado
        }
    }
    return 0;
}

int main() 
{
    int v[TAM];
    int n, elemento, resultado;

    scanf("%d", &n);
    lerVetor(n, v);
    scanf("%d", &elemento);

    // Realiza a busca binária e imprime o resultado
    resultado = PesqBin(n, elemento, v);
    if (resultado != 0) 
    {
        printf("%d\n", resultado);
    } 
    else 
    {
        printf("0\n");
    }

    printf("\n");
    return 0;
}

File: CI1056 - Algoritmos 2/Binário/BuscaBinária.pas
program buscabinaria;

type
    vetor = array [1..200] of integer;

var
    v : vetor;
    n, elemento: integer;

procedure lerVetor (n : integer; var v: vetor);
var
    i: integer;
begin

    for i:= 1 to n do
    begin
        read(v[i]);
    end;

end;

function PesqBin (n: integer; elemento: integer; var v: vetor):integer;
var
    meio, inicio, fim: integer;
begin
    inicio:= 1;
    fim:= n;
    repeat
        meio:= (inicio + fim) div 2;
        if(elemento > v[meio]) then
            inicio:= meio + 1
        else
            fim:= meio - 1;
    until (elemento = v[meio]) or (inicio > fim);
    if(elemento = v[meio]) then
        PesqBin:= meio
    else
        PesqBin:= 0;
end;

begin
    read(n);
    lerVetor(n, v);
    read(elemento);
    PesqBin(n, elemento, v);
    if(PesqBin(n, elemento, v) <> 0) then
        writeln(PesqBin(n, elemento, v))
    else
        writeln(PesqBin(n, elemento, v));
end.